File: 721-accounts-merge/accounts-merge.py
class DSU:

  def __init__(self, n: int):
    self.parent = list(range(n))

  def find(self, x: int) -> int:
    if self.parent[x] != x:
      self.parent[x] = self.find(self.parent[x])
    return self.parent[x]

  def union(self, x: int, y: int) -> None:
    root_x = self.find(x)
    root_y = self.find(y)
    if root_x != root_y:
      self.parent[root_x] = root_y


class Solution:

  def accountsMerge(self, accounts: list[list[str]]) -> list[list[str]]:
    n = len(accounts)
    dsu = DSU(n)
    email_to_id = {}

    for i, account in enumerate(accounts):
      for email in account[1:]:
        if email not in email_to_id:
          email_to_id[email] = i
        else:
          dsu.union(i, email_to_id[email])

    root_to_emails = {}
    for email, idx in email_to_id.items():
      root = dsu.find(idx)
      if root not in root_to_emails:
        root_to_emails[root] = []
      root_to_emails[root].append(email)

    result = []
    for idx, emails in root_to_emails.items():
      result.append([accounts[idx][0]] + sorted(emails))

    return result

File: 721-accounts-merge/accounts-merge.java
class Solution {
    static class DSU {
        int[] parent;

        DSU(int n) {
            parent = new int[n];
            for (int i = 0; i < n; i++) parent[i] = i;
        }

        int find(int x) {
            if (parent[x] == x) return x;
            return parent[x] = find(parent[x]);
        }

        void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            if (rootX != rootY) {
                parent[rootX] = rootY;
            }
        }
    }

    public List<List<String>> accountsMerge(List<List<String>> accounts) {
        int n = accounts.size();
        DSU dsu = new DSU(n);
        Map<String, Integer> emailToId = new HashMap<>();

        for (int i = 0; i < n; i++) {
            for (int j = 1; j < accounts.get(i).size(); j++) {
                String email = accounts.get(i).get(j);
                if (!emailToId.containsKey(email)) {
                    emailToId.put(email, i);
                } else {
                    dsu.union(i, emailToId.get(email));
                }
            }
        }

        Map<Integer, List<String>> rootToEmails = new HashMap<>();
        for (Map.Entry<String, Integer> entry : emailToId.entrySet()) {
            int root = dsu.find(entry.getValue());
            rootToEmails.computeIfAbsent(root, k -> new ArrayList<>()).add(entry.getKey());
        }

        List<List<String>> result = new ArrayList<>();
        for (Map.Entry<Integer, List<String>> entry : rootToEmails.entrySet()) {
            List<String> emails = entry.getValue();
            Collections.sort(emails);
            List<String> account = new ArrayList<>();
            account.add(accounts.get(entry.getKey()).get(0));
            account.addAll(emails);
            result.add(account);
        }

        return result;
    }
}